apps/ex_wire/config/test.exs

```diff
@@ -10,6 +10,7 @@ config :ex_wire,
     port: 30_399
   ],
   sync_mock: ExWire.BridgeSyncMock,
+  kademlia_mock: ExWire.FakeKademlia,
```

Comment:
I think this parameter shouldn't be called `kademlia_mock`. Maybe just `kademlia`? It will be mocked in test env

Reply:
The reason I named it this way was to keep consistency with the above.

Reply:
I think the idea is that in the non-test environments, the application will handle it properly -- it shouldn't be up to config. With config, you can choose to mock it, but otherwise, config shouldn't be specifying an implementation:
```
@kademlia Application.get_env(:ex_wire, :kademlia_mock, Kademlia)
```
Let me know if we agree to change this, though, because I use this pattern elsewhere and will update that as well.

Reply:
I do think this isn't the most elegant way to mock, because you can specify only one function behaviour per environment. One of the approaches that gives you full control of the behaviour is to always pass the dependencies between processes down the supervision tree, but it gives much overhead on implementation...
If it was me... I would start using Mock libraries. Like :meck for example. WDYT @ayrat555 @willmeister @hayesgm ?

Reply:
I *really* dislike Meck, and significantly prefer this variety (or argument-pass) behaviours. Meck changes the system under test too much that it makes it difficult to understand behavior. I generally like doing:

```
@my_lib Application.get_env(:ex_wire, :some_lib, DefaultLib)

def my_function(arg1, arg2, my_lib \\ @my_lib) do
   my_lib.some_function()
end
```

This makes it very easy to inject mock versions of a library in a way that is very clear and predictable. With a behavior on the library, you can also have dialyzer verify that the "mock" version matches the signature of the true library. Additionally, it enforces that you should only mock when you *truly* have a library that is only useful when it is side-effecting. The reduction in side-effects and push toward pure functions is a bonus.